content/blog/2024/09/usr-merge/index.md

```diff
@@ -0,0 +1,90 @@
+2) The Usr Merge work depended on a previous update already having been installed on user's systems. When fetching updates, `eopkg` goes straight from where you are to the latest package version found in the repository, skipping anything in between. Since the team's systems are usually always up-to-date on the Unstable repository, we didn't catch that if you weren't already up-to-date, you would be missing the necessary work done in a previous update when you tried to install the Usr Merge update.
+
+The second issue is compounded by the first issue because even if the second is solved, the first makes it impossible to reliably install an update addressing it. We were stuck in the position where some users were successfully Usr Merged, other users weren't, and there was no way to unify those two groups, either by successfully merging everyone, or rolling back the merge.
+
+We started calling it the Chicken and the Egg problem.
+
+## Chicken and the egg
+
+The problem goes like this: we can update `eopkg` to provide more deterministic updates. In fact, Ermo has already done this for both the [Python 2](https://github.com/getsolus/eopkg/pull/68) and [Python 3](https://github.com/getsolus/eopkg/pull/70) versions of `eopkg`. But because updates are already non-deterministic, and we can't guarantee what point in time users are updating from, we can't rely on the update being present on user systems if we try to do the Usr Merge again; trying to update would lead to the same problem as before.
+
+And that holds true for any fix that involves a package update to solve this problem. No matter what package it is, or what the update does, the issue is always the same.
+
+## Cracking the egg
+
+We have come up with a way forward that should solve our little Chicken and Egg problem. We can't rely on a single update in a single package to solve the issue, but what we *can* do is update the pieces of software involved to write and look for marker files signaling that they are compatible with Usr Merge, and each other. In conjunction, we can ship a script to be run during the early boot process to do the actual merging, if and only if the marker files are in place. By having the process rely on these files and boot scripts instead of just happening with update installation, we can ensure that the merging process will only run when all of the pieces are in place, and not before. The package manager is taken out of the equation entirely. If you want to see what the script does for yourself, it can be found [here](https://github.com/getsolus/packages/blob/main/packages/u/usysconf-epoch/files/epoch.sh).
+
+### The plan
+
+The process goes like this:
+
+1) eopkg is updated with Usr-Merge support. It will write a marker file to indicate that the system can safely be merged.
```

Comment:
```suggestion
1) eopkg is updated with Usr Merge support. It writes a marker file to indicate that the system can safely be merged.
```
Was written without hyphen up to here (and further down) [though personally I'm more a fan of the hyphenated version]